numpyro/contrib/control_flow/scan.py

```diff
@@ -87,45 +87,53 @@ def _subs_wrapper(subs_map, i, length, site):
                                " which is currently not supported. Please report the issue to us!")
 
 
-class promote_shapes(Messenger):
-    # a helper messenger to promote shapes of `fn` and `value`
-    #   + msg: fn.batch_shape = (2, 3), value.shape = (3,) + fn.event_shape
-    #     process_message(msg): promote value so that value.shape = (1, 3) + fn.event_shape
+class _promote_fn_shapes(Messenger):
+    # a helper messenger to promote shapes of `fn`
     #   + msg: fn.batch_shape = (3,), value.shape = (2, 3) + fn.event_shape
     #     process_message(msg): promote fn so that fn.batch_shape = (1, 3).
-    def process_message(self, msg):
+    def postprocess_message(self, msg):
         if msg["type"] == "sample" and msg["value"] is not None:
             fn, value = msg["fn"], msg["value"]
             value_batch_ndims = jnp.ndim(value) - fn.event_dim
             fn_batch_ndim = len(fn.batch_shape)
-            prepend_shapes = (1,) * abs(fn_batch_ndim - value_batch_ndims)
-            if fn_batch_ndim > value_batch_ndims:
-                msg["value"] = jnp.reshape(value, prepend_shapes + jnp.shape(value))
-            elif fn_batch_ndim < value_batch_ndims:
+            if fn_batch_ndim < value_batch_ndims:
+                prepend_shapes = (1,) * (value_batch_ndims - fn_batch_ndim)
                 msg["fn"] = tree_map(lambda x: jnp.reshape(x, prepend_shapes + jnp.shape(x)), fn)
 
 
-def scan_enum(f, init, xs, length, reverse, rng_key=None, substitute_stack=None):
+def _promote_scanned_value_shapes(value, fn):
+    # a helper function to promote shapes of `value`
+    #   + msg: fn.batch_shape = (T, 2, 3), value.shape = (T, 3,) + fn.event_shape
+    #     process_message(msg): promote value so that value.shape = (T, 1, 3) + fn.event_shape
+    value_batch_ndims = jnp.ndim(value) - fn.event_dim
+    fn_batch_ndim = len(fn.batch_shape)
+    if fn_batch_ndim > value_batch_ndims:
+        prepend_shapes = (1,) * (fn_batch_ndim - value_batch_ndims)
+        return jnp.reshape(value, jnp.shape(value)[:1] + prepend_shapes + jnp.shape(value)[1:])
+    else:
+        return value
+
+
+def scan_enum(f, init, xs, length, reverse, rng_key=None, substitute_stack=None, history=1,
+              first_available_dim=None):
     from numpyro.contrib.funsor import config_enumerate, enum, markov
     from numpyro.contrib.funsor import trace as packed_trace
 
-    # XXX: This implementation only works for history size=1 but can be
-    # extended to history size > 1 by running `f` `history_size` times
-    # for initialization. However, `sequential_sum_product` does not
-    # support history size > 1, so we skip supporting it here.
-    # Note that `funsor.sum_product.sarkka_bilmes_product` does support history > 1.
+    # amount number of steps to unroll
+    history = min(history, length)
+    unroll_steps = min(2 * history - 1, length)
```

Comment:
This is the minimum number of unroll_steps to get a correct last carry shape. But it is safe to use any number greater than this number.